makeflow/src/makeflow.c

```diff
@@ -172,6 +175,27 @@ static struct makeflow_alloc *storage_allocation = NULL;
 /* Variables used to hold the time used for storage alloc. */
 uint64_t static_analysis = 0;
 
+#define MAKEFLOW_HOOK_CALL(hook_name, ...) do { \
+	if (!makeflow_hooks) \
+		makeflow_hooks = list_create(); \
+	list_first_item(makeflow_hooks); \
+	for (struct makeflow_hook *h; (h = list_next_item(makeflow_hooks));) { \
+		int rc = MAKEFLOW_HOOK_SUCCESS; \
+		if (h->hook_name) \
+			rc = h->hook_name(__VA_ARGS__); \
+		if (rc !=MAKEFLOW_HOOK_SUCCESS) \
+			fatal("hook %s:" #hook_name " returned %d",h->module_name?h->module_name:"", rc); \
+	} \
+} while (0)
```

Comment:
Let's not fill up the main loop with macro expansions, that can lead to crazy error messages and difficult debugging.
In the main loop of makeflow, just make simple function calls to `makeflow_hook_x()` etc at the key points.
Then, define `makeflow_hook_x()` in `makeflow.c` and let that handle the iteration over the various hook objects.
You can use macros there if you like.